src/high-level/constructors.lisp

```diff
@@ -0,0 +1,155 @@
+                 type))
+           (tensor (make-tensor (list (floor range)) tensor-class element-type :order order))
+           (f (lambda (index)
+                (coerce index element-type))))
+      (specialize-tensor (into! f tensor)))))
+
+(defgeneric from-array (array shape &key type order)
+  (:documentation "Create a tensor from an array")
+  (:method (array shape &key type (order :row-major))
+    (let* ((element-type
+             (if (null type)
+                 (array-element-type array)
+                 type))
+           (tensor-class
+             (if (null type)
+                 (compatible-tensor-constructors element-type)
+                 (compatible-tensor-constructors type))))
+      (adjust-array array (list (reduce #'* shape)) :element-type element-type)
+      (specialize-tensor
+       (make-tensor shape tensor-class element-type
+                    :storage array
+                    :order order)))))
+
+(defgeneric from-list (list shape &key type order input-order)
+  (:documentation "Create a tensor of the specified shape from a list, putting elements in row-major order.
+NOTE: When type is not specified, the type is inferred from the first element of the list")
+  (:method (list shape &key type (order :column-major) (input-order :row-major))
+    (check-type shape shape)
+    (let ((shape-size (reduce #'* shape))
+          (list-size (length list)))
+      (assert (cl:= list-size shape-size)
+              () "Incompatible shape. Must have the same total number of elements. The list has ~a elements and the new shape has ~a elements" list-size shape-size))
+    (let* ((tensor-class
+             (if (null type)
+                 (compatible-tensor-constructors-from-value (first list))
+                 (compatible-tensor-constructors type)))
+           (element-type
+             (if (null type)
+                 (upgraded-array-element-type (type-of (first list)))
+                 type))
+           (tensor (make-tensor shape tensor-class element-type :order order)))
+      (specialize-tensor
+       (into!
+        (lambda (&rest pos)
+          (coerce (nth
+                   (if (eql input-order :row-major)
+                       (row-major-index pos shape)
+                       (column-major-index pos shape))
+                   list)
+                  element-type))
+        tensor)))))
+
+(defgeneric from-diag (list shape &key type order)
+  (:documentation "Create a tensor of the specified shape from a list, placing along the diagonal
+NOTE: When type is not specified, the type is inferred from the first element of the list")
+  (:method (list shape &key type (order :column-major))
+    (check-type shape shape)
+    (assert (cl:= 2 (length shape))
+            () "Shape must be of rank 2.")
+    (assert-square-shape shape)
+    (let ((shape-size (reduce #'* shape))
+          (list-size (length list)))
+      (assert (cl:= (* list-size list-size) shape-size)
```

Comment:
```suggestion
      (assert (cl:= list-size (first shape)))
```

already `assert`ed that `shape` is square and length 2. You could also skip the `shape-size` calculation if you tweak the error message below.